src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectContextFactory.cs

```diff
@@ -0,0 +1,88 @@
+using Microsoft.VisualStudio.LanguageServices.Implementation.TaskList;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    [Export(typeof(IProjectContextFactory))]
+    internal partial class ProjectContextFactory : IProjectContextFactory
+    {
+        private readonly VisualStudioProjectTracker _projectTracker;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly VisualStudioWorkspaceImpl _visualStudioWorkspace;
+        private readonly HostDiagnosticUpdateSource _hostDiagnosticUpdateSource;
+        private readonly Dictionary<string, IVsReportExternalErrors> _externalErrorReporterMap;
+        private readonly ImmutableDictionary<string, string> _projectLangaugeToErrorCodePrefixMap =
+            ImmutableDictionary.CreateRange(StringComparer.OrdinalIgnoreCase, new[]
+            {
+                new KeyValuePair<string, string> (LanguageNames.CSharp, "CS"),
+                new KeyValuePair<string, string> (LanguageNames.VisualBasic, "BC"),
+            });
+
+        [ImportingConstructor]
+        public ProjectContextFactory(
+            VisualStudioProjectTracker projectTracker,
+            SVsServiceProvider serviceProvider,
+            VisualStudioWorkspaceImpl visualStudioWorkspace,
+            HostDiagnosticUpdateSource hostDiagnosticUpdateSource)
+        {
+            _projectTracker = projectTracker;
+            _serviceProvider = serviceProvider;
+            _visualStudioWorkspace = visualStudioWorkspace;
+            _hostDiagnosticUpdateSource = hostDiagnosticUpdateSource;
+            _externalErrorReporterMap = new Dictionary<string, IVsReportExternalErrors>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        // internal for testing purposes only.
+        internal static IProjectContext CreateProjectShim(VisualStudioProjectTracker projectTracker, IServiceProvider serviceProvider, IVsHierarchy hierarchy, string projectDisplayName, string projectFilePath, string language, Guid projectGuid, string projectTypeGuid, CommandLineArguments commandLineArguments)
+        {
+            return new ProjectContext(commandLineArguments, projectTracker, reportExternalErrorCreatorOpt: null, hierarchy: hierarchy, language: language,
+                serviceProvider: serviceProvider, visualStudioWorkspaceOpt: null, hostDiagnosticUpdateSourceOpt: null, projectDisplayName: projectDisplayName,
+                projectFilePath: projectFilePath, projectGuid: projectGuid, projectTypeGuid: projectTypeGuid);
+        }
+
+        IProjectContext IProjectContextFactory.CreateProjectContext(string languageName, string projectDisplayName, string projectFilePath, Guid projectGuid, string projectTypeGuid, object hostObject, CommandLineArguments commandLineArguments)
+        {
+            var vsHierarchy = hostObject as IVsHierarchy;
```

Comment:
Indeed, I'd rather we declare our warts than try to hide them. Using 'object' might mean we can avoid a API break in the signature, but since the code on both sides is still strongly coupled on both sides. Also, as long as the RDT still talks in hierarchies as it's currency, I'm not sure we'll be able to remove that. I can think of some other API shapes which might make this more strongly typed, but I still forsee the parameter here being the least of our worries when such a refactoring would happen.